src/routeSpec/Routes.js

```diff
@@ -0,0 +1,60 @@
+'use strict'
+
+class Routes {
+  /**
+   * Declares all of the Route specifications for the application
+   * @param {Array.<RouteSpec>} routeSpecs
+   */
+  constructor (routeSpecs) {
+    this._routeSpecs = routeSpecs
+  }
+
+  /**
+   * Matches events with matching Route specifications.
+   * @param {Object} event An AWS event
+   * @returns {RouteSpec} If a match is found
+   * @throws {Error} If no match is found
+   */
+  matches (event) {
+    const pathMatches = this._routeSpecs.filter(spec => this._hasPath(event, spec))
+    if (pathMatches.length < 1) throw new Error('Not Found')
+
+    const methodMatches = pathMatches.filter(spec => this._hasMethod(event, spec))
+    if (methodMatches.length < 1) throw new Error('Method Not Allowed')
+
+    const pathParamsMatches = methodMatches.filter(spec => this._hasPathParams(event, spec))
+    if (pathParamsMatches.length < 1) throw new Error('Not Found')
+
+    const queryMatches = pathParamsMatches.filter(spec => this._hasQueryParams(event, spec))
+    if (queryMatches < 1) throw new Error('Bad Request')
+
+    return (queryMatches.length === 1) ? pathMatches[0] : (() => { throw new Error('Internal server error') })()
```

Comment:
Since you are writing every check with if, I would suggest adding the rule . Now you are mixing a  the last check with the return value and it is confusing because  it is not two return values based on a condition.
```
if(queryMatches.length !==1) {throw new Error(...)}

return pathMatches[0]
```